engine/uState.ml

```diff
@@ -24,6 +24,76 @@ type uinfo = {
   uloc : Loc.t option;
 }
 
+type quality = QVar of Sorts.QVar.t | QProp | QSProp | QType
+
+module QState : sig
+  type t
+  type elt = Sorts.QVar.t
+  val empty : t
+  val union : t -> t -> t
+  val add : elt -> t -> t
+  val repr : elt -> t -> quality
+  val set : elt -> quality -> t -> t option
+  val collapse : t -> t
+  val pr : t -> Pp.t
+end =
+struct
+
+module QMap = Map.Make(Sorts.QVar)
+
+type t = quality option QMap.t
+(* TODO: use a persistent union-find structure *)
+
+type elt = Sorts.QVar.t
+let empty = QMap.empty
+let union s1 s2 = QMap.fold QMap.add s1 s2
+let add q m = QMap.add q None m
+
+let rec repr q m = match QMap.find q m with
+| None -> QVar q
+| Some (QVar q) -> repr q m
+| Some (QProp | QSProp | QType as q) -> q
+| exception Not_found ->
+(*   let () = assert !Flags.in_debugger in *) (* FIXME *)
```

Comment:
Fixme?

Reply:
IIRC some code in ssr plays fast and loose with the ambient ustate for printing, leading to a breakage of this invariant. Nothing serious but will take ages to fix.